Quickstart Documentation + Refactor Sample Strategy (#224)
neume-network/strategies: docs/quickstart.md

# quickstart

> As described in the [README](https://github.com/neume-network/strategies/readme.md), a strategy requires an `extractor` and, optionally, a `transformer`:

Author: `or` or `and/or` @il3ven ?

> As described in the [README](https://github.com/neume-network/strategies/readme.md), a strategy requires an `extractor` and, optionally, a `transformer`:

Reviewer: ```suggestion
As described in the [README](https://github.com/neume-network/strategies/readme.md), a strategy requires an `extractor` or a `transformer` or both:
```

This should be more clear.

This guide will show you the steps to crawl XKCD.

First off, the `core` expects, both for `init` and `update`, a message compliant to the [schema definition for https strategies](https://github.com/neume-network/schema/blob/main/src/schema.mjs).

the `init` function message should provide the starting state of the crawling. For XKCD, we would write something like:
```
{
    write: null,
    messages: [
      {
        type: "https",
        commissioner: name,
        version,
        options: {
          url: https://xkcd.com/1/info.0.json,
          method: "GET",
          header: null,
          body: null
        },
        results: null,
        error: null,
      },
    ],
  };
```

* `write`: specifies the data to be written in string format. `null` during init, as there is no data
* `type`: the protocol. must be `https` for this type of strategy
* `commissioner`: ??
* `version`: can be anything, it's currently not used
* `options`:
  * `url`: the URL pointing to the first chunk to crawl. First page on XKCD is `1`
  * `method`: GET, POST
  * `headers`: if any specific header should be sent to the source, it shall be specified here (for instance authorization headers)

>   * `body`: same

Reviewer 2: ```suggestion
  * `body`: same, value must be type of `String`
```

* `results`: this field will be used by the core to pass the output of the crawling to the `update` function

Once the `init` message has been returned, the `core` will fetch the first chunk of data and pass it to `update`. 

> The function, ideally, should take care of the following:

Author: Isn't this a bit of a repetition of what is written right below @il3ven ? Or was your intention to have a more summarized description of the `update` role?

> The function, ideally, should take care of the following:

Reviewer: My main intention was to point out that update is similar to init. Instead of storing, it can also send new messages to worker. For an example you can see [soundxyz-metadata](https://github.com/neume-network/strategies/blob/main/src/strategies/soundxyz-metadata/extractor.mjs#L106).

> The function, ideally, should take care of the following:

Author: @il3ven i added what you requested, just a bit more consistent with the rest of the sentence. Let me know if it's ok

* validate the outcome of the crawl
* validate the data against a schema
* prepare the data for storing. This step should make sure that any future consumer of the data will find all that's required to process it.
* define if and what to crawl next

The `core` provided message will contain the following data:
```
{
    "error": string,
    "results": object
}
```

* error: null or the crawling error message
* results: the direct output of the crawler

> The strategy will have to

Reviewer 2: About the sections below suggesting to exit the entire crawl, I'd say this suggestion may not produce good outcomes. I'd prefer an inconsistent but complete crawl over the inverse.

So usually, in case of errors we log the message to improve debugging experience. And we try to deal with errors or validation failures in ways where we can make the best of the current crawl.

But certainly, there are cases where e.g. we could find that a continuing a crawl makes no sense anymore e.g. in case we expected a certain schema but then the data source is now returning another one.

@sirnicolaz does this help? Happy to assist with the formulation of the text, please let me know.

Author: Thanks for you comment. So are you saying that crawl should just log and continue upon error, unless it's a schema validation problem, for which case the crawl should halt?

Author: I changed the comment on the first part (mentioning that if possible, the crawling should go on). let me know if it's sufficient @TimDaub
